scripts/sandbox.js

```diff
@@ -15,26 +18,51 @@ class Sandbox {
     });
     this.drawMarking(mouseX, mouseY);
   }
+
+  static redo() {
+    const changes = sandbox.changes;
+
+    if(changes.length - 1 > 0) {
+      changes.splice(changes.length - 1, 1);
+    }
+  }
+
   static pushLevel() {
+    const changes = sandbox.changes;
```

Comment:
I don't understand what this method does. Can you mansplain it to me? :)

Reply:
It puts the sandbox level into our scenes.

Reply:
The thing I don't understand is that you rewrite the whole level with a "change".  Isn't a change just one action? For example if I add two blocks, won't this only load the last one?

Reply:
No. It's unclear, I know, but the changes contain the level, but saves the previous levels aswell. If you log the `changes` array it keeps the history of all levels. That creates the opportunity to go undo :)

Reply:
So the changes looks like this (approximately)?

```javascript
[
]
```

Reply:
Yes. And the current scene (if in sandbox-mode) will always be the last value of the `changes` variable.
I find two things to change with this though:

1. Change name! Maybe to like `history`?
2. Maybe the `sandbox.level` is not even needed since the level will always be the last value in the `changes` variable. That is why we `splice()` in the `redo()` method (soon to be called `undo()`).

Reply:
But I think you may have gotten the wrong order. We just use `push()`, which puts the new value in the end of the array.

Reply:
Why not save it like this?

```javascript
```

Reply:
Why would we?

Reply:
So we can easily undo just one cube.

```typescript
changes[0] = changes[0].slice(0, 1)
```

Reply:
That's pretty much what we're doing, just that it isn't 2 dimensions. We do:
`history.splice(history.length - 1, 1);`

Reply:
We are saving all added cubes, step by step, and go "back in time" when we undo.

Reply:
Now I get it. I looked at the variables when trying it myself. Interesting way of solving it! :)

My misunderstanding was that I thought each index was a "level"

Like this:
```javascript
[
"", // <- level 1 history.
"". // <- level 2 history.
]
```